aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -2650,18 +2660,48 @@ Tensor& linalg_matrix_norm_out(
 
 // Numerical or None norms
 Tensor linalg_norm(const Tensor& self, const optional<Scalar>& opt_ord, optional<IntArrayRef> opt_dim, bool keepdim, optional<ScalarType> opt_dtype) {
+  int64_t ndim = self.dim();
+  std::vector<int64_t> dim_ = opt_dim.has_value() ? opt_dim.value().vec() : make_dim_list(ndim);
+  if (!opt_ord.has_value() || dim_.size() == 1) {
+    return at::linalg_vector_norm(self, opt_ord.value_or(2), opt_dim, keepdim, opt_dtype);
+  }
+
   auto options = TensorOptions().dtype(opt_dtype.has_value() ? opt_dtype.value() : toRealValueType(self.scalar_type())).device(self.device());
   Tensor result = at::empty({0}, options);
-  return at::native::linalg_norm_out(
-      self, opt_ord, opt_dim, keepdim, opt_dtype, result);
+
+ if (dim_.size() == 2) {
+    _linalg_norm_matrix_out(result, self, opt_ord.value(), dim_, keepdim, opt_dtype);
+  } else {
+    TORCH_CHECK(false, "'dim' must specify 1 or 2 dimensions when order is numerical and input is "
+                "not 1-D or 2-D");
+  }
+
+ return result;
 }
 
 // Frobenius and nuclear norms
 Tensor linalg_norm(const Tensor& self, c10::string_view ord, optional<IntArrayRef> opt_dim, bool keepdim, optional<ScalarType> opt_dtype) {
+  check_str_ord_valid(ord, opt_dim, self.dim());
+
+  if (ord == "fro") {
+    return at::linalg_vector_norm(self, 2, opt_dim.value_or(IntArrayRef({0, 1})),
+                                  keepdim, opt_dtype);
+  }
+
   auto options = TensorOptions().dtype(opt_dtype.has_value() ? opt_dtype.value() : toRealValueType(self.scalar_type())).device(self.device());
   Tensor result = at::empty({0}, options);
-  return at::native::linalg_norm_out(
-      self, ord, opt_dim, keepdim, opt_dtype, result);
+  Tensor self_ = opt_dtype.has_value() ? self.to(opt_dtype.value()) : self;
+
+  // These _out functions need to be called since some of the tests expect the result to be of a particular dtype,
+  // which the non-out functions don't support.
+  if (opt_dim.has_value()) {
+    at::nuclear_norm_out(result, self_, opt_dim.value(), keepdim);
```

Comment:
This works almost by pure chance. This `_out` variant happens to be a composite function (_out variants should not be composite functions nor differentiable).
Could you leave a comment explaining this?

Reply:
The comment I was expecting went more along the lines of:
> // Even though these look like `_out` variants, they are in fact composite operations. While this is odd, it's fine to call them here, as they make `linalg_norm` still a composite operation.